### ISSUE TITLE
Handyman stuck in Toilet, blocking patient

### ISSUE THREAD

**Issue by maugsburger** [CorsixTH/CorsixTH#1105]
#### Describe the issue

There is a handyman stuck in the toilet room, blocking  the patient washing her hands from moving.

#### Steps to Reproduce
1. *Load save game attached to this ticket*

2. Try to remove patient or handyman from the room.

3. Game becomes unusable (eg pickup will block mouse forever)

#### Save Game
autosave6 is the latest before the problem, in autosave7 it already happened. The handyman right behind the woman on the toilet doesn't move, neither pickup directly nor from the staff details popup works.

I didn't manage to reproduce it by waiting after autosave6.

#### Expected Behavior

- The Handyman leaves the room
- I can fire the handyman, send the patient home, and either will move.

#### System Information
**CorsixTH Version:** 0.60

**Operating System:** Windows 7 64bit

**Theme Hospital Version:** GOG.com

#### Gamelog.txt
```
Created game with savegame version 111.
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\class.lua:136: Suspected infinite loop
stack traceback:
	[C]: in function 'error'
	C:\Program Files\CorsixTH\Lua\app.lua:919: in hook '?'
	C:\Program Files\CorsixTH\Lua\class.lua:136: in field 'is'
	C:\Program Files\CorsixTH\Lua\entity_map.lua:54: in method 'addEntity'
	C:\Program Files\CorsixTH\Lua\entity.lua:133: in field 'setTile'
	C:\Program Files\CorsixTH\Lua\entities\patient.lua:1057: in method 'setTile'
	C:\Program Files\CorsixTH\Lua\entity.lua:174: in method 'setTilePositionSpeed'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:176: in function <C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:137>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:580: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:672: in method 'finishAction'
	...ogram Files\CorsixTH\Lua\humanoid_actions\use_object.lua:325: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:219: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1075: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1000: in function <C:\Program Files\CorsixTH\Lua\app.lua:997>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:901: in function <C:\Program Files\CorsixTH\Lua\app.lua:896>
Recovering from error in timer handler...
Error in timer handler: 
C:\Program Files\CorsixTH\Lua\utility.lua:220: Suspected infinite loop
stack traceback:
	[C]: in function 'error'
	C:\Program Files\CorsixTH\Lua\app.lua:919: in hook '?'
	C:\Program Files\CorsixTH\Lua\utility.lua:220: in function 'ipairs'
	C:\Program Files\CorsixTH\Lua\entity_map.lua:72: in local 'remove_entity_from_table'
	C:\Program Files\CorsixTH\Lua\entity_map.lua:88: in method 'removeEntity'
	C:\Program Files\CorsixTH\Lua\entity.lua:119: in method 'setTile'
	C:\Program Files\CorsixTH\Lua\entity.lua:174: in method 'setTilePositionSpeed'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:176: in function <C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:137>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:580: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:656: in method 'queueAction'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\meander.lua:95: in field '?'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:580: in upvalue 'Humanoid_startAction'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:672: in method 'finishAction'
	C:\Program Files\CorsixTH\Lua\humanoid_actions\idle.lua:31: in local 'timer_function'
	C:\Program Files\CorsixTH\Lua\entity.lua:219: in field 'tick'
	C:\Program Files\CorsixTH\Lua\entities\staff.lua:149: in method 'tick'
	C:\Program Files\CorsixTH\Lua\world.lua:1075: in method 'onTick'
	C:\Program Files\CorsixTH\Lua\app.lua:1000: in function <C:\Program Files\CorsixTH\Lua\app.lua:997>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:901: in function <C:\Program Files\CorsixTH\Lua\app.lua:896>
Recovering from error in timer handler...
Savegame version is 111 (v0.60), originally it was 111 (v0.60)
Error in buttonup handler: 
C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:60: attempt to call a nil value (local 'timer_function')
stack traceback:
	C:\Program Files\CorsixTH\Lua\humanoid_actions\walk.lua:60: in local 'on_interrupt'
	C:\Program Files\CorsixTH\Lua\entities\humanoid.lua:643: in method 'setNextAction'
	C:\Program Files\CorsixTH\Lua\dialogs\staff_dialog.lua:317: in local 'callback'
	C:\Program Files\CorsixTH\Lua\window.lua:682: in method 'handleClick'
	C:\Program Files\CorsixTH\Lua\window.lua:1478: in field 'onMouseUp'
	C:\Program Files\CorsixTH\Lua\dialogs\staff_dialog.lua:273: in method 'onMouseUp'
	C:\Program Files\CorsixTH\Lua\window.lua:1462: in field 'onMouseUp'
	C:\Program Files\CorsixTH\Lua\ui.lua:649: in function <C:\Program Files\CorsixTH\Lua\ui.lua:636>
	(...tail calls...)
	C:\Program Files\CorsixTH\Lua\app.lua:901: in function <C:\Program Files\CorsixTH\Lua\app.lua:896>
```

### Additional Info
The savegames can be found at https://www.dropbox.com/sh/xq1x3m45v92s8pk/AAA2uh7I0DJZBRH2BI6VAKJXa?dl=0

![image](https://cloud.githubusercontent.com/assets/1416183/16466125/4534852a-3e42-11e6-8a8a-a88463afb80c.png)

**Comment by Brainsample** [comment]
Hi Maugsburger,

Thanks for your report. Running your "autosave6.sav" some strange things can be seen. A handyman sometimes walks through the wall (the tile just to the right of the door) into the toiletroom. Then sometimes the handyman tries to walk the same way back through the wall. This causes either a crash or a "Warning: Humanoid leaving a room that they are not in"

The handymen don't have anything to do in the toiletroom, they just seem to meander in and out. The patient in the toiletroom never moves, despite having an action queue.

When you place a plant outside the toiletroom (blocking the path through the wall) the game doesn't crash, but the toiletroom remains unusable.

**Comment by lewri** [comment]
Do you have an autosave of before the woman enters the toiletroom? Looks like something has broken earlier than autosave 6, which could contribute to the handyman walking through the wall.

I did manage to stop the error but picking up the handyman and taking him out the room before using the door to exit. Looks like it freezes as soon as the handyman tries to exit in correct fashion. Will continue investigating.

**Comment by maugsburger** [comment]
> Do you have an autosave of before the woman enters the toiletroom? Looks like something has broken earlier than autosave 6, which could contribute to the handyman walking through the wall.

Yes, I added the additional save files into the Dropbox folder. The order according to my local timestamps is 11-12-1-2-...

**Comment by nickstar1234** [comment]
try not to keep the toilet to close to the sink, when i build a rest room, i tend to leave a gap between sink and the toilet. hope that helps

**Comment by mugmuggy** [comment]
Just double checking this one after looking at #1094. I can reproduce the circumstances with the saves, and when loading Autosave6 the issue has already occured but the cause was right there. Autosave5 has at (52,74) objects a bin and a radiator.  Move the radiator and the travelWest flag is reset to true on (53,74) (and on 52,74 travelEast), effectively creating a path through the wall. Loading Autosave6 I assume that the radiator we've picked up is the one from the (52,74) tile.

What I haven't been able to do is replicate it away from this save, but could replicate the same issue at (48,80)(49,80), where on (49,80) is another radiator and bin but this doesn't have a negative affect on path finding it seems.

**Comment by TheCycoONE** [comment]
@mugmuggy you're saying this is a duplicate of #1094?

**Comment by mugmuggy** [comment]
Appears so. Wall tile flags are punching a hole in the walls with both instances. I'm speculating on #1094 as I don't have before save but in that one there is a radiator placed a little weirdly, which could have been moved away from the wall resulting in the same issue.

**Comment by TheCycoONE** [comment]
Link to mugmuggy's insight: https://github.com/CorsixTH/CorsixTH/issues/1094#issuecomment-337849879

**Comment by mugmuggy** [comment]
Can replicate it now. Place radiator first, build room up to radiator, move radiator away from wall.

Believe the issue llies in Object:setTile. It calls the local function setPassableFlags, only for side objects, which updates the flags either side on remove. When the object is placed against the wall after the room is built, the placement logic skips calling setPassableFlags as the flag is already false'

`          if xy.only_side then`
`            if map:getCellFlags(lx, ly)[passable_flag] == true then`
`              self.set_passable_flags = true`
`              setPassableFlags(passable_flag, lx, ly, next_tile_x, next_tile_y, false)`

Something like this in setPassableFlags, will detect the wall boundary and prevent the issue from occuring. with the infinite loop, due to paths being cut through walls.

`    local room_pos = self.world.map:getRoomId(xpos, ypos)`
`    local room_next = self.world.map:getRoomId(next_x, next_y)`
`    if room_pos ~= room_next then`
`      return`
`    end`

But this also leaves any back to back side objects, also becoming completely passable if one object is removed, allowing humanoids to pass straight through the remaining object.

Here I blocked off a path. Don't think this should be possible, but path finding sends them through the other door.
![blocked](https://user-images.githubusercontent.com/8009292/32001203-75473b88-b9dc-11e7-8004-be9b5b33b7a9.png)

And here I removed one of the radiators and captured a patient walking through the radiator.
![unblocked](https://user-images.githubusercontent.com/8009292/32001259-9b4a1292-b9dc-11e7-85b7-4e4083b856be.png)

So there is this deeper issue, but this doesn't cause an infinite loop or anything

**Comment by TheCycoONE** [comment]
Thanks @mugmuggy 

I don't think the room check is the right approach there, but an object check might be.

**Comment by mugmuggy** [comment]
By the looks of it, we can conditionally if so desired call
`  self.world.map.th:updatePathfinding()`
`  self.world:resetSideObjects()`
in Entity:onDestroy after the setTile

My understanding is updatePathfinding appears to rebuild the broken wall flags that setTile(nil) breaks in setPassableFlags

resetSideObjects fixes the back to back side objects becoming a path leak, if the first placed one is removed.
